bluepysnap/query.py

```diff
@@ -19,12 +21,57 @@
 ALL_KEYS = {NODE_ID_KEY, EDGE_ID_KEY, POPULATION_KEY, OR_KEY, AND_KEY, NODE_SET_KEY} | VALUE_KEYS
 
 
-# TODO: move to `libsonata` library
+def _logical_and(masks):
+    """Calculate and return the "logical and" of the given masks.
+
+    Args:
+        masks: list of array-like, or booleans.
+
+    Returns:
+        the resulting mask as a numpy array, or a bool.
+        Return True if the masks list is empty, similarly to all([]).
+    """
+    filtered_masks = []
+    for mask in masks:
+        if mask is False:
+            return False
+        if mask is True:
+            continue
+        filtered_masks.append(mask)
+    if not filtered_masks:
+        return True
+    # reduce(operator.and_, masks) is faster than np.logical_and.reduce() or np.all()
+    return reduce(operator.and_, (np.asarray(m, dtype=bool) for m in filtered_masks))
+
+
+def _logical_or(masks):
+    """Calculate and return the "logical or" of the given masks.
+
+    Args:
+        masks: list of array-like, or booleans.
+
+    Returns:
+        the resulting mask as a numpy array, or a bool.
+        Return False if the masks list is empty, similarly to any([]).
+    """
+    filtered_masks = []
+    for mask in masks:
+        if mask is True:
+            return True
+        if mask is False:
+            continue
+        filtered_masks.append(mask)
+    if not filtered_masks:
+        return False
+    # reduce(operator.or_, masks) is faster than np.logical_or.reduce() or np.any()
+    return reduce(operator.or_, (np.asarray(m, dtype=bool) for m in filtered_masks))
```

Comment:
Wow, this is lightning fast, tried a few things but just couldn't beat this. Good job!